R/validate-policy.R

```diff
@@ -1,26 +1,35 @@
 #' Validate Policy File
 #' 
 #' Flattens and validates policy file
-#'  @param policy YAML policy file
-#'  @param gs Gradescope data
-#'  @importFrom purrr map discard
-#'  @export
+#' @param policy YAML policy file
+#' @param gs Gradescope data
+#' @importFrom purrr map discard
+#' @export
 validate_policy <- function(policy, gs){
   policy <- flatten_policy(policy)
-  # drop categories with unavailable assignments
```

Comment:
Coming back to the question of speed, If this function is anything but non-trivial in terms of run time on Stat 20, can you use a line-profiling tool to figure out where the bottlenecks are? There's a great discussion of that here:

https://adv-r.hadley.nz/perf-measure.html?q=profiling#perf-measure

Reply:
I'll look through that link, but I'm pretty sure the biggest waste of time is having to iterate through different levels of nesting because each iteration only goes through one level of nesting:
So, assume gs doesn't have any Lab-related assignments (think week 2 of Stat 20):
First iteration: drops all Lab-related assignments in policy + Lab 1, Lab 2, etc because they're null
Second iteration: drops Lab 1, Lab 2 from Labs and then drops Labs because it's null
Third iteration: drop Labs from Overall Grade
It's really intensive because it keeps iterating until policy$categories stops changing in length, so this is 100% a slow placeholder for when we find a more efficient way to run this

Reply:
Having to repeat a significant amount of work each iteration is likely the biggest factor into the computational complexity. That said, it is still only linear in the deepest nested unvalid category, which is not too bad. I see a few places in the code where the speed can be improved. We should test it, but I doubt it takes very long and it should not have to be called that often.

Reply:
Nikita, you bring up a good point: it might be best to trim some (s)lowhanging fruit if there are any, but otherwise keep this as a placeholder to be revisited for speed. The line profiling would probably best be done on the full set of validate and calculate. The whole philosophy of line profiling is to do it empirically and try to identify: where are we spending, say, 90% of our time? and then aim to shrink that, then move on to the nextx 90% item.